### ISSUE TITLE
UDL syntax highlighting and folding regions are completely broken.

### ISSUE THREAD

**Issue by Darxide23** [notepad-plus-plus/notepad-plus-plus#10007]
### Description of the Issue
I'm trying to use a UDL for AutoHotkey and it's just simply not working out quite right. Others on the AHK forums where this UDL was posted have the same issue and they have just resorted to a work around of deselecting the language and then reselecting the AHK language. This may fix the highlighting problems, but it does not fix the problems with folding regions.

The highlighting problems is so far only related to comments. At random, Npp will believe large swaths (or sometimes the entire script) is a comment and highlight it as such. AHK has both inline and block commenting and Npp seems to interpret some inline comments as block comments despite inline and block comment deliminators being completely different.

As for the folding regions, I cannot see any rhyme or reason as to how Npp is inserting these. They are in a constant state of flux. Adding lines, removing lines, or sometimes just editing lines will cause these folding regions to jump all over the place. Sometimes I can make edits to a file, observe the folding regions change, then undo the edits with Ctrl+Z and see the folding regions change to something completely different to what they were before.

### Steps to Reproduce the Issue
The problem with folding regions is always present and doesn't seem like it can be fixed at all. The comment problem occurs at random. Unfortunately I cannot suggest any steps to reproduce this as I have no idea what triggers it. But again, as for the folding region problems, those are just always present.

### Debug Information
Notepad++ v7.9.5 (64-bit)
Build time : Mar 21 2021 - 02:13:17
Path : D:\program files\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Windows 10 Pro (64-bit)
OS Version : 1903
OS Build : 18362.30
Current ANSI codepage : 1252
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

I have also tried this with v8.0 with identical results.

The UDL in question may be viewed here: https://pastebin.com/cKYrW5ex

Attached are two screenshot examples of the bonkers folding regions.

![1623677593686-comments](https://user-images.githubusercontent.com/41874525/121939054-2b084600-cd12-11eb-8f15-9950ce16a775.png)

![Untitled](https://user-images.githubusercontent.com/41874525/121939063-2e033680-cd12-11eb-906e-9c548747bdc7.png)

**Comment by pryrt** [comment]
@Darxide23 ,

Can you attach the file, rather than just linking to the pastebin?  The developers like issues to be self-contained, and don't want to have to go download files from some other site.  This may improve the chances that it will be fixed.

To attach, you can just drag-and-drop the file into your reply to attach it.)

-- @pryrt (aka [PeterJones](https://community.notepad-plus-plus.org/user/peterjones) from the Forum)

**Comment by ludamo** [comment]
I had a quick look at your AHK xml file and found that it had a '%' as the 8th Group - prefix mode - Keyword list tab as well as having '%' in the Operators 1 section of the Operators & Delimiters tab. If I might suggest removing one of those '%' to see if it helps?

**Comment by Darxide23** [comment]
> I had a quick look at your AHK xml file and found that it had a '%' as the 8th Group - prefix mode - Keyword list tab as well as having '%' in the Operators 1 section of the Operators & Delimiters tab. If I might suggest removing one of those '%' to see if it helps?

That didn't help, however I unchecked the box allowing folding for comments and the problems just all vanish instantly.

**Comment by ludamo** [comment]
Yes on looking more closely at your AHK code I see that you only had pairs of '%@@%' and not the '% @@' which is allowed in AHK and which caused problems for syntax highlighting in the past. If you haven't looked at the following 2 series of posts they may provide some other hints. It can be frustrating!
https://community.notepad-plus-plus.org/topic/21326/what-is-going-on-with-this-user-defined-language
https://community.notepad-plus-plus.org/topic/20755/syntax-highlighting-works-for-a-time-then-irreproducibly-bleeds-only-restarting-npp-fixes-it?_=1623900006876

**Comment by chcg** [comment]
There seems to be a different UDL for AHK available at https://github.com/notepad-plus-plus/userDefinedLanguages/blob/master/UDLs/AutoHotKey_byBrotherGabriel-Marie.xml.

Probably the issue should be reported at https://github.com/jNizM/ahk_notepad-plus-plus.

Maybe creating a PR if the versions from @jNizM at https://github.com/notepad-plus-plus/userDefinedLanguages would be nice.

**Comment by Darxide23** [comment]
> https://community.notepad-plus-plus.org/topic/21326/what-is-going-on-with-this-user-defined-language

That is my post.

As for the other, I've already tried adding ((EOL)) and it didn't seem to help. but the comment highlighting usually doesn't happen right away. It only happens after Npp has been open for a while (days).

**Comment by pryrt** [comment]
@Darxide23 ,

> The author is aware of the problem

Which of the users in the thread is this "author" you are talking about?  

None of the people posting in https://community.notepad-plus-plus.org/topic/21326/what-is-going-on-with-this-user-defined-language use the jNizM handle.  None of the posts I have seen in that thread claim to be by the author of the UDL.

The only one in that forum discussion to mention a workaround of switching to Normal Text and back to the UDL was astrofista who is _not_ (as far as I know) the same person as jNizM.

So I'm not sure why you think that you've sufficiently requested that the author of the UDL look into it.  (Unless it's at some other discussion that you haven't linked to.)

**Comment by Darxide23** [comment]
The amount of incredulity is not encouraging.

The author is aware:
https://www.autohotkey.com/boards/viewtopic.php?f=88&t=50&start=120#p328150

**Comment by pryrt** [comment]
> This topic was linked all the way in my original post.

But it wasn't linked anywhere in this issue.  As the person requesting that something be fixed, it's generally considered polite to make the request self-contained, so that people trying to help don't have to go digging through a series of 

> The amount of incredulity is not encouraging.

I'm sorry that I'm not very encouraging, but I didn't post here in order to encourage you.  I'm doing this because I want to increase the chances that this will be fixed (*), and that is best done by making all the details as obvious as possible and easily accessible to someone reading this official issue.  Since the information wasn't here, I was challenging you to bring it all together in this one location.  Because if the people who do have the knowledge and authority to make codebase changes come here and they cannot find the information right away, the chances are much higher that they won't prioritize this request.

*: I especially want this because it's not just the auto-hot-key UDL.  I have seen other UDL that do not consistently apply the code folding, so that some times they fold fine, but other times the folding appears to be in random places -- but I've never been able to get one to misbehave reliably enough to make a feature request.  Since the AHK UDL does seem to trigger this more reliably (at least for you), I was hoping it would be the testbed going forward.  Or, if one of the folks here with more knowledge of the codebase was able to find what kind of discrepancies in the AHK UDL definition are causing the folding to get out of whack, then I could use that knowledge to go back and debug the definitions in my UDL and see if they have the same type of problems.

**Comment by jNizM** [comment]
UDL is updated for N++ v8.1.
Please test and report bugs or submit a change request.
